File: src/optic/lua/flat/ops/ForRangeLoop.java
package optic.lua.flat.ops;

import optic.lua.flat.*;

import java.util.List;
import java.util.stream.*;

public class ForRangeLoop implements Step {
	private final StepType type;
	private final String varName;
	private final Register from;
	private final Register to;
	private final List<Step> block;

	public ForRangeLoop(StepType type, String varName, Register from, Register to, List<Step> block) {
		this.type = type;
		this.varName = varName;
		this.from = from;
		this.to = to;
		this.block = List.copyOf(block);
	}

	@Override
	public StepType getType() {
		return type;
	}

	@Override
	public String toString() {
		return typeName() + " " + varName + " = " + from + ", " + to;
	}

	@Override
	public Stream<Step> children() {
		return block.stream();
	}
}

File: src/optic/lua/asm/instructions/ForRangeLoop.kt
package optic.lua.asm.instructions

import optic.lua.asm.*
import java.util.function.Consumer

class ForRangeLoop(
        val counter: VariableInfo,
        val from: RValue,
        val to: RValue,
        val block: AsmBlock) : Step {
    override fun <T : Any, X : Throwable> accept(visitor: StepVisitor<T, X>): T = visitor.visitForRangeLoop(counter, from, to, block)


    override fun toString(): String {
        return "for $counter = $from, $to"
    }

    override fun children(): List<Step> {
        return block.steps()
    }

    override fun forEachObserved(action: Consumer<RValue>) {
        action.accept(from)
        action.accept(to)
    }
}
